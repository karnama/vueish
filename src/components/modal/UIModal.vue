<template>
    <teleport to="body">
        <div v-if="isOpen"
             class="h-full w-full z-40 fixed top-0 left-0 wrapper"
             :class="{ 'visible': isVisible }">
            <!--Modal overlay-->
            <div class="overlay absolute h-screen w-screen" @click="close('cancel')" />

            <!--Main modal dialog-->
            <UIPanel class="modal top-0 w-full overflow-y-scroll
                            relative bg-default m-auto shadow-xl rounded bg-white opacity-0 translate-y-0"
                     role="dialog"
                     aria-modal="true"
                     :aria-label="header"
                     v-bind="$attrs"
                     no-collapse>
                <!--Modal Header-->
                <template #header>
                    <slot name="header">
                        {{ header }}
                    </slot>
                </template>

                <!--Modal Close Icon-->
                <template #actions>
                    <slot name="actions">
                        <i class="fa fa-times cursor-pointer text-muted
                    hover:text-default transition-text-color show-on-hover"
                           @click="close('cancel')">close</i>
                    </slot>
                </template>

                <!--Modal Body-->
                <slot>
                    {{ body }}
                </slot>

                <!--Modal Footer-->
                <template #footer>
                    <slot name="footer" :close="close">
                        <UIButton class="ui-modal-close ml-auto mr-2" @click="close('cancel')">
                            {{ closeButtonLabel }}
                        </UIButton>
                        <UIButton category="primary" class="ui-modal-accept" @click="close('accept')">
                            {{ acceptButtonLabel }}
                        </UIButton>
                    </slot>
                </template>
            </UIPanel>
        </div>
    </teleport>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue';
import UIButton from '@components/button/UIButton.vue';
import UIPanel from '@components/panel/UIPanel.vue';
import type { PropType } from 'vue';

export default defineComponent({
    name: 'UIModal',

    components: {
        UIButton,
        UIPanel
    },

    inheritAttrs: false,

    props: {
        /**
         * Header string
         *
         * @type {String|Object}
         */
        header: {
            type: String
        },

        /**
         * Body string
         *
         * @type {String|Object}
         */
        body: {
            type: String
        },

        /**
         * Callback to execute on modal cancel.
         */
        closeCallback: {
            type: Function as PropType<(callback: () => Promise<void>) => Promise<void>>
        },

        /**
         * The label for the close button.
         *
         * @type {String|Object}
         */
        closeButtonLabel: {
            type: String,
            default: 'Cancel'
        },

        /**
         * The label for the accept button.
         */
        acceptButtonLabel: {
            type: String,
            default: 'Confirm'
        }
    },

    emits: ['cancel', 'accept'],

    setup(props, ctx) {
        const isOpen = ref(false);
        const isVisible = ref(false);

        const open = async (): Promise<void> => {
            isOpen.value = true;
            return new Promise(resolve => setTimeout(() => {
                isVisible.value = true;
                resolve();
            }, 100));
        };

        const close = async (event: 'accept' | 'cancel' = 'cancel'): Promise<void> => {
            // filter out events if user doesn't define the argument
            event = ['accept', 'cancel'].includes(event) ? event : 'cancel';

            const closeModal = async (): Promise<void> => {
                isVisible.value = false;
                ctx.emit(event);
                return new Promise(resolve => setTimeout(() => {
                    isOpen.value = false;
                    resolve();
                }, 100));
            };

            if (typeof props.closeCallback === 'function') {
                return props.closeCallback(closeModal);
            }

            return closeModal();
        };

        return {
            isOpen,
            isVisible,
            open,
            close
        };
    }
});
</script>

<style scoped lang="scss">
.overlay {
    background: #00000061;
}

.wrapper {
    opacity: 0;
    transition: opacity 200ms linear;
    overflow: hidden;
}

.modal {
    max-width: 700px;
    max-height: 80%;
    transition: transform 300ms ease-out, opacity 100ms linear;
}

.wrapper.visible {
    opacity: 1;

    .modal {
        transform: translateY(70px);
        opacity: 1;
    }
}
</style>
