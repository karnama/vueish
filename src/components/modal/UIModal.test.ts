import UIModal from './UIModal.vue';
import { mount } from '@vue/test-utils';

describe('UIModal', () => {
    it('should renders', () => {
        const wrapper = mount(UIModal);
        expect(wrapper).toBeDefined();
        expect(wrapper.exists()).toBe(true);
        wrapper.unmount();
    });

    it('should display the actions slot content',  async () => {
        const wrapper = mount(UIModal, {
            slots: {
                actions: '<div id="test-actions" />'
            }
        });

        // @ts-expect-error
        await wrapper.vm.open();

        expect(document.getElementById('test-actions')).not.toBeNull();
        wrapper.unmount();
    });

    it('should emit closed on click outside or cancel click', async () => {
        const wrapper = mount(UIModal);

        // @ts-expect-error
        await wrapper.vm.open();
        // @ts-expect-error
        await document.querySelector('.overlay')?.click();

        expect(wrapper.emitted()['cancel']).toHaveLength(1);
        wrapper.unmount();
    });

    it('should emit accepted on accept click', async () => {
        const wrapper = mount(UIModal, {
            props: {
                acceptButtonLabel: 'acceptBtn'
            }
        });

        // @ts-expect-error
        await wrapper.vm.open();
        let button: HTMLButtonElement | undefined;
        document.querySelectorAll('button').forEach(btn => {
            if (btn.innerHTML === 'acceptBtn') {
                button = btn;
            }
        });

        if (button) {
            button.click();
        } else {
            throw new Error('Button not found in DOM');
        }

        expect(wrapper.emitted()['accept']).toHaveLength(1);
        wrapper.unmount();
    });

    it('should execute callbacks if given', async () => {
        const func = jest.fn();
        const wrapper = mount(UIModal, {
            props: {
                acceptButtonLabel: 'acceptBtn',
                accept: func
            }
        });

        // @ts-expect-error
        await wrapper.vm.open();
        let button: HTMLButtonElement | undefined;
        document.querySelectorAll('button').forEach(btn => {
            if (btn.innerHTML === 'acceptBtn') {
                button = btn;
            }
        });

        if (button) {
            button.click();
        } else {
            throw new Error('Button not found in DOM');
        }

        expect(func).toHaveBeenCalled();
        wrapper.unmount();
    });

    it('should display the labels defined by the prop', async () => {
        const wrapper = mount(UIModal, {
            props: {
                acceptButtonLabel: 'acceptBtn',
                closeButtonLabel: 'closeBtn'
            }
        });

        // @ts-expect-error
        await wrapper.vm.open();
        let count = 0;
        document.querySelectorAll('button').forEach(button => {
            if (button.innerHTML === 'acceptBtn' || button.innerHTML === 'closeBtn') {
                count++;
            }
        });
        expect(count).toBe(2);
        wrapper.unmount();
    });

    it('should display the header slot content', async () => {
        const wrapper = mount(UIModal, {
            slots: {
                header: '<div id="header-content" />'
            }
        });

        // @ts-expect-error
        await wrapper.vm.open();
        expect(document.getElementById('header-content')).not.toBeNull();
        wrapper.unmount();
    });

    it('should display the footer slot content', async () => {
        const wrapper = mount(UIModal, {
            slots: {
                footer: '<div id="footer-content" />'
            }
        });

        // @ts-expect-error
        await wrapper.vm.open();
        expect(document.getElementById('footer-content')).not.toBeNull();
        wrapper.unmount();
    });

    it('should display the header prop content', async () => {
        const wrapper = mount(UIModal, {
            props: {
                header: 'header-content'
            }
        });

        // @ts-expect-error
        await wrapper.vm.open();
        expect(document.querySelector('.modal')?.innerHTML).toContain('header-content');
        wrapper.unmount();
    });

    it('should display the default slot content', async () => {
        const wrapper = mount(UIModal, {
            slots: {
                default: '<div id="default-content" />'
            }
        });

        // @ts-expect-error
        await wrapper.vm.open();
        expect(document.getElementById('default-content')).not.toBeNull();
        wrapper.unmount();
    });

    it('should open and close using methods through the reference', async () => {
        const wrapper = mount(UIModal, {
            slots: {
                default: '<div id="default-content" />'
            }
        });

        expect(document.getElementById('default-content')).toBeNull();
        // @ts-expect-error
        await wrapper.vm.open();
        expect(document.getElementById('default-content')).not.toBeNull();

        // @ts-expect-error
        await wrapper.vm.close();
        expect(document.getElementById('default-content')).toBeNull();
        wrapper.unmount();
    });
});
