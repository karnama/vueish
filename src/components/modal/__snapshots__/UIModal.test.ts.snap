// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIModal should render correctly 1`] = `
<div
  class="h-full w-full z-40 fixed top-0 left-0 wrapper visible"
>
  <!--Modal overlay-->
  <div
    class="overlay absolute h-screen w-screen"
  />
  <!--Main modal dialog-->
  <section
    aria-expanded="true"
    aria-modal="true"
    class="ui-panel rounded shadow-md relative transition-all
                    bg-white dark:bg-gray-600 text-color modal top-0 w-full overflow-y-scroll
                            relative bg-default mx-auto shadow-xl rounded bg-white opacity-0 translate-y-0"
    role="dialog"
  >
    <transition-stub>
      
      
    </transition-stub>
    <header
      class="w-full flex items-center px-12 relative transition-all no-blur"
    >
      <h1
        class="select-none flex-grow py-8"
      >
        
        
        Header slot content
        
        
      </h1>
      <div>
        
        
        Actions slot content
        
        
      </div>
    </header>
    <transition-stub>
      
      <div
        class="no-blur"
      >
        <main
          class="w-full px-12 py-6"
        >
          
          
          Default slot content
          
          
        </main>
        <footer
          class="py-8 px-12 flex items-center"
        >
          
          
          <button
            class="border-gray-300 text-gray-900 bg-gray-300 disabled:bg-gray-300 ring-gray-600 hover:bg-gray-400 hover:border-gray-400 border px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed ui-modal-close ml-auto mr-2"
            type="button"
          >
            <transition-stub>
              
              <span
                class="label"
              >
                
                Cancel
                
              </span>
              
            </transition-stub>
          </button>
          <button
            class="border-blue-600 text-white bg-blue-600 ring-blue-300 hover:bg-blue-700 hover:border-blue-700 border px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed ui-modal-accept"
            type="button"
          >
            <transition-stub>
              
              <span
                class="label"
              >
                
                Confirm
                
              </span>
              
            </transition-stub>
          </button>
          
          
        </footer>
      </div>
      
    </transition-stub>
    <transition-stub>
      
      
    </transition-stub>
  </section>
</div>
`;
